### PR TITLE
RzIL TriCore Support

### PR DESCRIPTION
**Pull request by b1llow**
**tc1.6.2**

https://www.infineon.com/dgdl/Infineon-AURIX_TC3xx_Architecture_vol1-UserManual-v01_00-EN.pdf?fileId=5546d46276fb756a01771bc4c2e33bdd
https://www.infineon.com/dgdl/Infineon-AURIX_TC3xx_Architecture_vol2-UserManual-v01_00-EN.pdf?fileId=5546d46276fb756a01771bc4a6d73b70

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

**Test plan**

CI is green

### CODE REVIEW

File: librz/analysis/arch/tricore/tricore_il.inc
Comment (XVilka) [inline librz/analysis/arch/tricore/tricore_il.inc]: Note, that PCXI structure depends on the Tricore version used: https://gitlab.com/qemu-project/qemu/-/issues/1453

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/arch/tricore/tricore_il.inc
Comment (b1llow) [reply librz/analysis/arch/tricore/tricore_il.inc]: > Note, that PCXI structure depends on the Tricore version used: [gitlab.com/qemu-project/qemu/-/issues/1453](https://gitlab.com/qemu-project/qemu/-/issues/1453)

Do we need to support all versions, or just a limited few?

Comment (XVilka) [reply librz/analysis/arch/tricore/tricore_il.inc]: @imbillow all. Also, it might be necessary for the future ISA versions as well, since they might change something in the future again. Thus, we should build the ISA versioning in.  See, for example, how it's handled in `librz/analysis/p/analysis_mips_cs.c` (search for `analysis->cpu`).

Comment (XVilka) [reply librz/analysis/arch/tricore/tricore_il.inc]: For testing purposes, unlike in the code, it's probably better to limit ourselves to the two latest major ISA versions though - 1.3.1 and 1.6.2

**Comment by XVilka** [comment]
@imbillow for FPU see RzILFloat:
- https://github.com/rizinorg/rizin/blob/dev/librz/il/theory_fbasic.c
- https://github.com/rizinorg/rizin/blob/dev/librz/il/definitions/float.c
- https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_il/rz_il_opcodes.h

Some are probably still missing, e.g. no square root yet for `QSEED.F` instruction. @Heersin please take a look at this PR too (`ftohp` instruction in particular):
> Convert the contents of data register D[a] from IEEE-754-2008 32-bit single precision floating point to IEEE-754-2008
16-bit half precision (data interchange) floating point format. The rounded result is put in data register D[c][15:0].
D[c][31:16] is set to zero.

**Comment by Heersin** [comment]
ok, will review it soon

File: librz/analysis/arch/tricore/tricore_il.inc
Comment (Heersin) [inline librz/analysis/arch/tricore/tricore_il.inc]: It would be helpful to add comments about the converting algorithm

**Review by Heersin — COMMENTED** [review]
Looks good.

File: librz/analysis/arch/tricore/tricore_il.inc
Comment (b1llow) [reply librz/analysis/arch/tricore/tricore_il.inc]: Look here <https://www.infineon.com/dgdl/Infineon-AURIX_TC3xx_Architecture_vol2-UserManual-v01_00-EN.pdf?fileId=5546d46276fb756a01771bc4a6d73b70>

Basically the pseudo code is given in the documentation

**Comment by XVilka** [comment]
@imbillow don't forget to rebase this one too, please, since the disassembly and analysis PR diverged a lot since

**Comment by XVilka** [comment]
@imbillow please also rebase this one. For now priority is DWARF and there is no need to focus on that meanwhile, but having only RzIL-related changes in this PR would still help.

**Comment by XVilka** [comment]
@imbillow now that capstone 5 PR was merged, please switch the base to `dev` and rebase again

**Comment by XVilka** [comment]
@imbillow this "masking broken tests" isn't necessary anymore

**Comment by XVilka** [comment]
1) Please rebase
2) I recommend to do this in parallel with DWARF though - e.g. one instruction at a time/day, this way the distraction from DWARF work will be minimal.

**Comment by b1llow** [comment]
The three failing tests should be resolved after updating the version of capstone-next. https://github.com/rizinorg/rizin/pull/4011

**Comment by XVilka** [comment]
@imbillow, I recommend extracting non-related changes into separate PRs once you finish implementing those. Currently, I see these could be separated and merged before:
- IL helpers change
- Tricore analysis plugin and disasm changes without RzIL functions

**Review by XVilka — CHANGES_REQUESTED** [review]
Before trace-testing, I recommend doing the "emulateme" style test first.
Once it's done, and you think all opcodes are implemented and CI is green, please split this PR into 3 parts:
1. Extract Tricore changes to RzAsm and RzAnalysis without RzIL changes into a separate PR, we will merge it first.
2. Extract RzIL additions (helpers, etc) that are not specific to Tricore into a separate PR; we will merge it before as well.
3. Keep only Tricore RzIL changes in this PR, we will merge it last, after first two merged and this one rebased on top of these.

**Comment by b1llow** [comment]
> Before trace-testing, I recommend doing the "emulateme" style test first. Once it's done, and you think all opcodes are implemented and CI is green, please split this PR into 3 parts:
> 
> 1. Extract Tricore changes to RzAsm and RzAnalysis without RzIL changes into a separate PR, we will merge it first.
> 2. Extract RzIL additions (helpers, etc) that are not specific to Tricore into a separate PR; we will merge it before as well.
> 3. Keep only Tricore RzIL changes in this PR, we will merge it last, after first two merged and this one rebased on top of these.

When I try to lift the float command, I realize that I want to specify the rounding mode at runtime, but the existing float in rzil doesn't have this capability. Should I leave the float part alone or change some of rzil's float modules to support this?

image

**Comment by XVilka** [comment]
@imbillow I suggest skipping floats for now, since both @Heersin and @DMaroo have some changes regarding these, lets make sure all general instructions work, then floats can be added in a separate PR.

**Review by XVilka — CHANGES_REQUESTED** [review]
```
 [XX] db/asm/tricore <asm> cmp.f d0, d0, d0
-- <asm> cmp.f d0, d0, d0 <--- 4b000000 ---> <IL>
-- IL
--- expected
+++ actual
@@ -1,1 +1,1 @@
-(set PSW (| (& (var PSW) (bv 32 0xffffff)) (<< (& (| (& (& (>> (var PSW) (bv 32 0x18) false) (bv 32 0xff)) (~ (& (>> (var d0) (bv 32 0x8) false) (bv 32 0xff)))) (& (& (>> (var d0) (bv 32 0x0) false) (bv 32 0xff)) (& (>> (var d0) (bv 32 0x8) false) (bv 32 0xff)))) (bv 32 0xff)) (bv 32 0x18) false)))
+(set d0 (| (ite (&& (! (|| (is_nan (var d0)) (is_nan (var d0)))) (<. (var d0) (var d0))) (bv 32 0x1) (bv 32 0x0)) (| (<< (ite (&& (! (|| (is_nan (var d0)) (is_nan (var d0)))) (! (|| (|| (is_nan (var d0)) (is_nan (var d0))) (|| (<. (var d0) (var d0)) (<. (var d0) (var d0)))))) (bv 32 0x1) (bv 32 0x0)) (bv 32 0x1) false) (| (<< (ite (&& (! (|| (is_nan (var d0)) (is_nan (var d0)))) (! (&& (! (|| (is_nan (var d0)) (is_nan (var d0)))) (|| (&& (! (|| (is_nan (var d0)) (is_nan (var d0)))) (<. (var d0) (var d0))) (&& (! (|| (is_nan (var d0)) (is_nan (var d0)))) (! (|| (|| (is_nan (var d0)) (is_nan (var d0))) (|| (<. (var d0) (var d0)) (<. (var d0) (var d0)))))))))) (bv 32 0x1) (bv 32 0x0)) (bv 32 0x2) false) (| (<< (ite (|| (is_nan (var d0)) (is_nan (var d0))) (bv 32 0x1) (bv 32 0x0)) (bv 32 0x3) false) (| (<< (ite (&& (is_zero (& (>> (var d0) (bv 32 0x17) false) (bv 32 0xff))) (! (is_zero (& (>> (var d0) (bv 32 0x0) false) (bv 32 0x7fffff))))) (bv 32 0x1) (bv 32 0x0)) (bv 32 0x4) false) (<< (ite (&& (is_zero (& (>> (var d0) (bv 32 0x17) false) (bv 32 0xff))) (! (is_zero (& (>> (var d0) (bv 32 0x0) false) (bv 32 0x7fffff))))) (bv 32 0x1) (bv 32 0x0)) (bv 32 0x5) false)))))))

WARNING: (../librz/il/il_validate.c:562):validate_pure_float_bool_uop: code should not be reached
IL Validation failed:
operand of is_nan op is not a float.
```

**Comment by DMaroo** [comment]
> When I try to lift the float command, I realize that I want to specify the rounding mode at runtime, but the existing float in rzil doesn't have this capability. Should I leave the float part alone or change some of rzil's float modules to support this?

It is not possible to do it in RzIL for the moment. And I don't think it would be a good idea either, since the format of a float is the part of its type and should be statically known in the IL. A workaround for this (and what I also did in x86 floating-point instructions lifting) is to use the `ITE` op and manually compare the value of a bitvector with all the permitted values and use the correct rounding mode in each of that branches.

An example macro (not linking the x86 code since that PR is yet to be merged):

```c
/**
 * \brief Execute the function \p f with the correct op mode argument
 *
 * \param f function which takes in the rounding mode as the first argument
 *
 * 0 -> RNE
 * 1 -> RTN
 * 2 -> RTP
 * 3 -> RTZ 
 */
#define EXEC_WITH_RMODE(f, ...) \
	ITE(EQ(VARL("_rmode"), UN(2, 0)), f(RZ_FLOAT_RMODE_RNE, __VA_ARGS__), \
		(EQ(VARL("_rmode"), UN(2, 1)), f(RZ_FLOAT_RMODE_RTN, __VA_ARGS__), \
			(EQ(VARL("_rmode"), UN(2, 2)), f(RZ_FLOAT_RMODE_RTP, __VA_ARGS__), \
				(f(RZ_FLOAT_RMODE_RTZ, __VA_ARGS__)))))

/* Need this because we want a function which takes in the rounding mode as the first argument. */
RzILOpFloat *convert_floating_helper(RzFloatRMode rmode, RzFloatFormat format, RzILOpFloat *val) {
	return FCONVERT(format, rmode, val);
}

void example(RzILOpPure *x, RzILOpPure *y, RzFloatFormat fmt) {
	/* We need to use DUP here so that the macro does not use reuse the same pure value everywhere. Not using DUP will lead to double-free. */
	RzILOpPure *_rounded_convert = EXEC_WITH_RMODE(convert_floating_helper, format, DUP(x));
	RzILOpPure *_rounded_add = EXEC_WITH_RMODE(FADD, DUP(x), DUP(y);
}
```

**Review by XVilka — CHANGES_REQUESTED** [review]
Ok, looks good now. I wonder if it makes sense for you to wait until TCG plugin API is implemented (for tracetesting), just like @dmaroo wants. Meanwhile, please split PRs as I suggested, and let's merge some of these.

**Comment by b1llow** [comment]
- https://github.com/rizinorg/rizin/pull/4092
- https://github.com/rizinorg/rizin/pull/4093

**Comment by b1llow** [comment]
> Ok, looks good now. I wonder if it makes sense for you to wait until TCG plugin API is implemented (for tracetesting), just like @DMaroo wants. Meanwhile, please split PRs as I suggested, and let's merge some of these.



> TCG plugin API is implemented

No problem, I can wait for the implementation of the TCG plugin API.

**Comment by XVilka** [comment]
@imbillow merged those, please rebase this one and mark ready for review

**Comment by XVilka** [comment]
@imbillow, I found few missing/problematic opcodes:
```
ℤ rizin -e asm.arch=tricore =
[0x00000000]> wx ef4f0800
[0x00000000]> pd 1
        ╭─< 0x00000000      ef4f0800       jz.t  d15, 0x14, 0x10
[0x00000000]> aoi
0x0 ()
[0x00000000]> ao
address: 0x0
opcode: jz.t d15, #0x14, #0x10
esilcost: 0
disasm: jz.t d15, #0x14, #0x10
mnemonic: jz.t
description: jump if zero bit
mask: ff000000
prefix: 0
id: 159
bytes: ef4f0800
refptr: 0
size: 4
sign: true
type: cjmp
cycles: 0
jump: 0x00000010
fail: 0x00000004
cond: eq
family: cpu
[0x00000000]> wx ef3fd17f
[0x00000000]> pd 1
        ╭─< 0x00000000      ef3fd17f       jz.t  d15, 0x13, 0xffffffa2
[0x00000000]> aoi
0x0 ()
[0x00000000]>
```
Also
```
ℤ rizin -e asm.arch=tricore =
 -- If you want to open the file in read-write mode, invoke rizin with '-w'
[0x00000000]> wx 90f2
[0x00000000]> pd 1
            0x00000000      90f2           addsc.a a2, a15, d15, #2
[0x00000000]> aoi
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
0x0 ()
[0x00000000]>
```

Moreover, when I tried `aaa; plf @@F` on some files, there were many warnings like that (can't share them, sorry):
```
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: Failed to get operand0 [2]: "ld.bu d15, #0" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: Failed to get operand0 [2]: "ld.hu d3, #0xa" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: Failed to get operand0 [2]: "ld.bu d15, #0" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: Failed to get operand0 [2]: "ld.q d6, #0x1450" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: Failed to get operand0 [2]: "ld.q d2, #0x2407" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: Failed to get operand0 [2]: "ld.q d3, #0x3406" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
WARNING: Failed to get operand0 [2]: "ld.q d6, #0x400000c0" [mem]
WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
```

**Comment by b1llow** [comment]
> @imbillow, I found few missing/problematic opcodes:
> 
> ```
> ℤ rizin -e asm.arch=tricore =
> [0x00000000]> wx ef4f0800
> [0x00000000]> pd 1
>         ╭─< 0x00000000      ef4f0800       jz.t  d15, 0x14, 0x10
> [0x00000000]> aoi
> 0x0 ()
> [0x00000000]> ao
> address: 0x0
> opcode: jz.t d15, #0x14, #0x10
> esilcost: 0
> disasm: jz.t d15, #0x14, #0x10
> mnemonic: jz.t
> description: jump if zero bit
> mask: ff000000
> prefix: 0
> id: 159
> bytes: ef4f0800
> refptr: 0
> size: 4
> sign: true
> type: cjmp
> cycles: 0
> jump: 0x00000010
> fail: 0x00000004
> cond: eq
> family: cpu
> [0x00000000]> wx ef3fd17f
> [0x00000000]> pd 1
>         ╭─< 0x00000000      ef3fd17f       jz.t  d15, 0x13, 0xffffffa2
> [0x00000000]> aoi
> 0x0 ()
> [0x00000000]>
> ```
> 
> Also
> 
> ```
> ℤ rizin -e asm.arch=tricore =
>  -- If you want to open the file in read-write mode, invoke rizin with '-w'
> [0x00000000]> wx 90f2
> [0x00000000]> pd 1
>             0x00000000      90f2           addsc.a a2, a15, d15, #2
> [0x00000000]> aoi
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> 0x0 ()
> [0x00000000]>
> ```
> 
> Moreover, when I tried `aaa; plf @@F` on some files, there were many warnings like that (can't share them, sorry):
> 
> ```
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: Failed to get operand0 [2]: "ld.bu d15, #0" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: Failed to get operand0 [2]: "ld.hu d3, #0xa" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: Failed to get operand0 [2]: "ld.bu d15, #0" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: Failed to get operand0 [2]: "ld.q d6, #0x1450" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: Failed to get operand0 [2]: "ld.q d2, #0x2407" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: Failed to get operand0 [2]: "ld.q d3, #0x3406" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> WARNING: (../librz/analysis/arch/tricore/tricore_il.c:3623):tricore_il_op: code should not be reached
> WARNING: Failed to get operand0 [2]: "ld.q d6, #0x400000c0" [mem]
> WARNING: (../librz/asm/arch/../../asm/arch/tricore/tricore.inc:100):tricore_op_as_mem: code should not be reached
> WARNING: SETG_wrap: assertion 'name && x && strlen(name) >= 2' failed (line 440)
> ```

The main reason is that I only referenced tc1.6.2, so other versions of the instructions may be a bit problematic.

**Comment by XVilka** [comment]
@imbillow strange, because those instructions are in the binary compiled for the tc1.6.2-compatible SoC.

**Comment by b1llow** [comment]
> @imbillow strange, because those instructions are in the binary compiled for the tc1.6.2-compatible SoC.

strange， I can't seem to find a description of the `ef4f0800 jz.t d15, #0x14, #0x10` instruction in either the tricore manuals or qemu. In capstone I remember referencing some version of qemu, but I can't seem to find it now.

- v1.6 https://www.infineon.com/dgdl/tc1_6__architecture_vol2.pdf?fileId=db3a3043372d5cc801374ad9c0653ad9
- TC1.6P & TC1.6E https://www.infineon.com/dgdl/Infineon-TC2xx_Architecture_vol1-UM-v01_00-EN.pdf?fileId=5546d46269bda8df0169ca1bea3624a5
- TC1.31 https://www.infineon.com/dgdl/tc_v131_instructionset_v138.pdf?fileId=db3a304412b407950112b409b6dd0352
- qemu https://gitlab.com/qemu-project/qemu/-/blob/master/target/tricore/tricore-opcodes.h

**Comment by b1llow** [comment]
> > @imbillow strange, because those instructions are in the binary compiled for the tc1.6.2-compatible SoC.
> 
> strange， I can't seem to find a description of the `ef4f0800 jz.t d15, #0x14, #0x10` instruction in either the tricore manuals or qemu. In capstone I remember referencing some version of qemu, but I can't seem to find it now.
> 
> * v1.6 https://www.infineon.com/dgdl/tc1_6__architecture_vol2.pdf?fileId=db3a3043372d5cc801374ad9c0653ad9
> * TC1.6P & TC1.6E https://www.infineon.com/dgdl/Infineon-TC2xx_Architecture_vol1-UM-v01_00-EN.pdf?fileId=5546d46269bda8df0169ca1bea3624a5
> * TC1.31 https://www.infineon.com/dgdl/tc_v131_instructionset_v138.pdf?fileId=db3a304412b407950112b409b6dd0352
> * qemu https://gitlab.com/qemu-project/qemu/-/blob/master/target/tricore/tricore-opcodes.h

Oh my bad.
I just checked byte[0] but the instruction OPCODE1 for BRN format is 0-6 bits.

**Review by XVilka — APPROVED** [review]
OK, now it's in quite good condition. Since we postponed trace-testing after TCG APIs were implemented, let's merge it as is. It works quite well already on most files I tried. It can be improved further incrementally, when necessary, and also during the trace testing.

**Pull request merged**
